main.rb

```diff
@@ -0,0 +1,115 @@
+# rubocop: disable Metrics/PerceivedComplexity, Metrics/CyclomaticComplexity
+module Enumerable
+  def my_each
+    x = 0
+    while x < length
+      yield(self[x])
+      x += 1
+    end
+  end
+
+  def my_each_with_index
+    i = 0
+    while i < length
+      yield(self[i], i)
+      i += 1
+    end
+    self
+  end
```

Comment:
## my_each_with_index

- [ ] returns an enumerator if no block is given

```Ruby
array.my_each_with_index
```
_expected : Enumerator
got: LocalJumpError:
       no block given (yield)_

- [ ] when a block is given when self is a range it returns the range itself after calling the given block once for each element in self

```Ruby
range.my_each_with_index(&block)
```

_got NameError:
       undefined local variable or method `length' for 5..50:Range_

- [ ] when a block is given when self is a range it calls the given block once for each element in self

```Ruby
block = proc { |num| my_each_output += num.to_s }
range.my_each_with_index(&block)
```

_got NameError:
       undefined local variable or method `length' for 5..50:Range_

- [ ] when a block is given when self is a hash it calls the given block once for each element in self

```Ruby
my_each_with_index_output = ''
block = proc { |num, idx| my_each_with_index_output += "Num: #{num}, idx: #{idx}\n" }
hash.my_each_with_index(&block)
```

_expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"
            got: "Num: , idx: 0\nNum: , idx: 1\nNum: , idx: 2\nNum: , idx: 3\nNum: , idx: 4\n"_